crates/ruff_python_formatter/src/expression/expr_dict.rs

```diff
@@ -1,21 +1,99 @@
-use crate::comments::Comments;
+use crate::comments::{dangling_node_comments, Comments};
+use crate::context::PyFormatContext;
 use crate::expression::parentheses::{
     default_expression_needs_parentheses, NeedsParentheses, Parentheses, Parenthesize,
 };
-use crate::{not_yet_implemented_custom_text, FormatNodeRule, PyFormatter};
+use crate::prelude::*;
+use crate::trivia::Token;
+use crate::trivia::{first_non_trivia_token, TokenKind};
+use crate::USE_MAGIC_TRAILING_COMMA;
+use crate::{FormatNodeRule, PyFormatter};
+use ruff_formatter::format_args;
 use ruff_formatter::{write, Buffer, FormatResult};
-use rustpython_parser::ast::ExprDict;
+use ruff_python_ast::prelude::Ranged;
+use rustpython_parser::ast::{Expr, ExprDict};
 
 #[derive(Default)]
 pub struct FormatExprDict;
 
+struct KvPair<'a> {
+    key: &'a Option<Expr>,
+    value: &'a Expr,
+}
+
+impl Format<PyFormatContext<'_>> for KvPair<'_> {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        if let Some(key) = self.key {
+            write!(
+                f,
+                [group(&format_args![
+                    key.format(),
+                    text(":"),
+                    space(),
+                    self.value.format()
+                ])]
+            )
+        } else {
+            write!(f, [group(&format_args![text("**"), self.value.format()])])
+        }
+    }
```

Comment:
Not a strong preference but it seems that we always want to group the whole item regardless of whether a key is present or not. Lifting the `group` out of the `KeyValuePair` formatting makes it clear that the grouping doesn't depend on the key's presence and it simplifies your formatting code (no need to use `format_args`). 

Hmm, yeah, using `group(&KeyValuePair { key, value })` probably won't work because of lifetimes. But the following should work

```rust
let joined = format_with(|f| {
			let mut joiner =  f.join_with(format_args!(text(","), soft_line_break_or_space()));

			for (key, value) in keys.iter().zip(values) {
				joiner.entry(group(&KeyValuePair { key, value }));
			}

             joiner.finish()
        });
```

Reply:
ah, makes sense for always grouping 👍 

(though to handle the leading comment this is no longer the case i think)